### ISSUE TITLE
Support for attribute type narrowing in structural pattern matching.

### ISSUE THREAD

**Issue by LeoCx1000** [microsoft/pyright#6649]
**Is your feature request related to a problem? Please describe.**
Yes, structural pattern matching for class patterns does not type narrow class attributes properly.

**Describe the solution you’d like**
The structural pattern matching would narrow the types based on what is given.
```python3
# pyright: strict
from typing import Any, reveal_type


class A:
    ...
class B:
    ...
class C:
    ...


class X:
    def __init__(self, thing: A | B | C):
        self.a_or_b: A | B | C = thing


for thing in (X(thing=A()), X(thing=B()), X(thing=C())):
    match thing:
        case X(a_or_b=A()):            # At type checking:
            reveal_type(thing.a_or_b)  # expected = A, actual = A | B | C
            print('expects A:', thing, thing.a_or_b)
        case X(a_or_b=B()):
            reveal_type(thing.a_or_b)  # expected = B, actual = A | B | C
            print('expects B:', thing, thing.a_or_b)
        case X(a_or_b=_):
            reveal_type(thing.a_or_b)  # expected = actual = A | B | C
            print('expects other:', thing, thing.a_or_b)

```

current solution is to write the instance checks redundantly, which gets cumbersome when doing longer case statements.

e.g.
```python3
case X(a_or_b=A()) if isinstance(thing.a_or_b, A):

```
*no idea how easy or hard this recursive type narrowing would be to implement, seems trivial at a glance but I ain't an expert ':(*

**Comment by erictraut** [comment]
During type narrowing, the narrowed type of an expression must be a valid type in the Python type system. There's no way in the Python type system to express the concept of "an instance of `X` whose attribute `a_or_b` has been narrowed to type `A`".

In your example above, the type of `thing` remains `X` regardless of which of these patterns it matches against, so `thing.a_or_b` remains `A | B | C`. 

If you want to narrow the type of `a_or_b`, you could make that the subject of the `match` statement, like this:

```python
for thing in (X(thing=A()), X(thing=B()), X(thing=C())):
    match thing.a_or_b:
        case A():
            reveal_type(thing.a_or_b) # A
            print("expects A:", thing, thing.a_or_b)
        case B():
            reveal_type(thing.a_or_b) # B
            print("expects B:", thing, thing.a_or_b)
        case C():
            reveal_type(thing.a_or_b) # C
            print("expects other:", thing, thing.a_or_b)
```

Alternatively, you could make `X` generic and use a constrained type variable, like this:

```python
class X[T: (A, B, C)]:
    def __init__(self, thing: T):
        self.a_or_b = thing

for thing in (X(thing=A()), X(thing=B()), X(thing=C())):
    match thing:
        case X(a_or_b=A()):
            reveal_type(thing.a_or_b) # A
            print('expects A:', thing, thing.a_or_b)
        case X(a_or_b=B()):
            reveal_type(thing.a_or_b) # B
            print('expects B:', thing, thing.a_or_b)
        case X(a_or_b=_):
            reveal_type(thing.a_or_b) # C
            print('expects other:', thing, thing.a_or_b)
```

This works because the type can be narrowed to `X[A]`, `X[B]` or `X[C]`.

**Comment by LeoCx1000** [comment]
Ohh okay. Thanks!
